bark_ml/observers/graph_observer.py

```diff
@@ -0,0 +1,554 @@
+      "dy": source_features["y"] - target_features["y"],
+      "dvel": source_features["vel"] - target_features["vel"],
+      "dtheta": source_features["theta"] - target_features["theta"]
+    }
+
+    features = {key: features[key] for key in self.enabled_edge_attribute_keys}
+    features = list(features.values())
+    assert len(features) == len(self.enabled_edge_attribute_keys)
+
+    return features
+
+  def _filter_requested_attributes(self, 
+                                   requested_keys, 
+                                   available_keys, 
+                                   context=""):
+    """
+    Filters the requested node/edge attributes by the available 
+    attributes and returns the intersection of the requested
+    and available attributes.
+    """
+    if not isinstance(requested_keys, list):
+      raise ValueError(
+        f"Requested {context} feature list must be a list of " +
+        f"strings, not {type(requested_keys)}.")
+
+    if len(requested_keys) == 0:
+      self._logger.warning(
+        "The `GraphObserver` received an empty list of requested " +
+        f"{context} attributes.")
+      return []
+
+    valid_keys = []
+    invalid_keys = []
+
+    for key in requested_keys:
+      if key in available_keys:
+        valid_keys.append(key)
+      else:
+        invalid_keys.append(key)
+    
+    if len(invalid_keys) > 0:
+      self._logger.warning(
+        f"The following {context} attributes requested from the GraphObserver " +
+        f"are not supported and will be ignored: {invalid_keys}")
+
+    return valid_keys
+
+  def _normalize_value(self, value, range):
+    """
+    Normalizes the `value` with can take on values in
+    into the range of [-1, 1].
+    If the `value` is outside the given range, it's clamped 
+    to the bound of [-1, 1]
+    """
+    normed = 2 * (value - range[0]) / (range[1] - range[0]) - 1
+    normed = max(-1, normed) # values lower -1 clipped
+    normed = min(1, normed) # values bigger 1 clipped
+    return normed
+
+  def reset(self, world):
+    return world
+
+  def _position(self, agent) -> Point2d:
+    return Point2d(
+      agent.state[int(StateDefinition.X_POSITION)],
+      agent.state[int(StateDefinition.Y_POSITION)]
+    )
+
+  @property
+  def normalization_data(self) -> Dict[str, list]:
+    """
+    The reference ranges of how certain attributes are normalized.
+    Use this info to scale normalized values back to real values.
+    
+    E.g. the value for key 'x' returns the possible range of 
+    the x-element of positions. A normalized value of 1.0 
+    corresponds to the maximum of this range (and vice versa).
+
+    Note: This dictionary does not include a range for each
+    possible attribute, but rather for each _kind_ of attribute,
+    like distances, velocities, angles, etc.
+    E.g. all distances (between agents, between objects, etc.)
+    are scaled relative to the 'distance' range. 
+    """
+    d = OrderedDict()
```

Comment:
what is the advantage of using an ordered dict?

Reply:
actually, there is none.

Reply:
that's what I wanted to know. Just curious; )

Reply:
I removed it for simplicity :)